nifi-nar-bundles/nifi-prometheus-bundle/nifi-prometheus-processors/src/main/java/org/apache/nifi/processors/prometheus/PrometheusRemoteWrite.java

```diff
@@ -0,0 +1,318 @@
+        serverEndpoint = new Server(threadPool);
+
+        ServerConnector connector = new ServerConnector(serverEndpoint);
+        connector.setPort(port);
+        serverEndpoint.addConnector(connector);
+
+        // Setup only one handler serving one context path
+        ContextHandler contextHandler = new ContextHandler();
+        contextHandler.setContextPath(contextPath);
+        contextHandler.setAllowNullPathInfo(true);
+        contextHandler.setHandler(new PrometheusHandler(context, session, maxBatch));
+        serverEndpoint.setHandler(contextHandler);
+
+        try {
+            serverEndpoint.start();
+            serverEndpoint.join();
+        } catch (Exception e) {
+            context.yield();
+            e.printStackTrace();
+        }
+    }
+
+    private class PrometheusHandler extends AbstractHandler {
+        private final JsonFormat.Printer JSON_PRINTER = JsonFormat.printer();
+        private ProcessSession session;
+        private ProcessContext context;
+        private FlowFile flowfile;
+        private int maxBatch;
+        private List<Metrics> metricList = new ArrayList<>();
+
+        public PrometheusHandler(ProcessContext context, ProcessSession session, int maxBatch) {
+            super();
+            this.session = session;
+            this.context = context;
+            this.maxBatch = maxBatch;
+        }
+
+        public void commitFlowFile(FlowFile flowFile, ProcessSession session, HttpServletRequest request){
+            session.putAttribute(flowfile, "mime.type","application/json");
+            session.transfer(flowfile, REL_SUCCESS);
+            session.getProvenanceReporter().receive(flowfile, request.getRequestURI());
+            session.commit();
+            getLogger().debug("SUCCESS relation FlowFile: {}.", new Object[]{flowfile.getId()});
+        }
+
+        @Override
+        public void handle(String target,
+                           Request baseRequest,
+                           HttpServletRequest request,
+                           HttpServletResponse response) throws IOException,ServletException {
+
+            // Uncompress the request on the fly: Snappy compressed protocol buffer message.
+            try (SnappyInputStream is = new SnappyInputStream(baseRequest.getInputStream())) {
+
+                if (is == null) {
+                     getLogger().error("InputStream is null");
+                }
+
+                ServletOutputStream responseOut = response.getOutputStream();
+                response.setStatus(HttpServletResponse.SC_OK);
+
+                WriteRequest writeRequest = WriteRequest.parseFrom(is);
+
+                for (Types.TimeSeries timeSeries: writeRequest.getTimeseriesList()) {
+                    List<MetricLabel> labelsList = new ArrayList<>();
+                    List<MetricSample> sampleList = new ArrayList<>();
+                    Metrics metrics = new Metrics();
+                    Gson gson = new Gson();
```

Comment:
IMO it would be better to use a RecordWriter (specified by the user) for output rather than only JSON, that way the user can decide which format (JSON, Avro, CSV, XML, etc.) to output the metrics in.